lib/src/commit_builder.rs

```diff
@@ -352,19 +351,14 @@ impl DetachedCommitBuilder {
     pub fn write(self, mut_repo: &mut MutableRepo) -> BackendResult<Commit> {
         let predecessors = self.commit.predecessors.clone();
         let commit = write_to_store(&self.store, self.commit, &self.sign_settings)?;
-        // FIXME: Google's index.has_id() always returns true.
-        if mut_repo.is_backed_by_default_index() && mut_repo.index().has_id(commit.id()) {
-            // Recording existing commit as new would create cycle in
-            // predecessors/parent mappings within the current transaction, and
-            // in predecessors graph globally.
-            return Err(BackendError::Other(
-                format!("Newly-created commit {id} already exists", id = commit.id()).into(),
-            ));
-        }
         mut_repo.add_head(&commit)?;
-        mut_repo.set_predecessors(commit.id().clone(), predecessors);
+        if !predecessors.contains(commit.id()) {
```

Comment:
nit: I would use `commit.id() != rewrite_source.id()` for consistency. If the caller set weird predecessors, I think panicking is good.